TrackletAlgorithm/VMSMERouter.h

```diff
@@ -0,0 +1,210 @@
+
+	// Number of bits for the memory bins
+	constexpr int nbitsbin = (layer) ? MEBinsBits : MEBinsBits + 1; // ME in disks has double the amount of bins
+
+	// Set values to VMStub
+	stub.setBend(bend);
+	stub.setIndex(index);
+	stub.setFinePhi(iphivmFineBins<inType>(phicorr, nbitsall + vmbits, nbitsfinephi));
+
+	// Indices used to find the rzfine value in LUT table
+	// LUT table returns the top 6 bits of a corrected z
+	// Note: not the index that is being saved to the stub
+	ap_uint<nbitsztable + nbitsrtable> indexz = (z >> (nbitsz - nbitsztable)); // Make z unsigned and take the top "nbitsztable" bits
+	ap_uint<nbitsrtable> indexr = -1;
+
+	constexpr int rbins = (1 << nbitsrtable); // Number of bins in r in LUT table
+
+	if (disk) {
+		if (negDisk) {
+			indexz = (1 << nbitsztable) -1 - indexz;
+		}
+		indexr = r;
+		if (inType == DISKPS) {
+			indexr = r >> (nbitsr - nbitsrtable); // Take the top "nbitsrtable" bits
+		}
+	} else { // layer
+		indexr = (r >> (nbitsr- nbitsrtable));// Make r unsigned and take the top "nbitsrtable" bits
+	}
+
+	// The index for LUT table
+	ap_uint<nbitsztable + nbitsrtable> lutIndex = (indexz * rbins) + indexr;
+
+	// Get the LUT value
+	auto lutValue = lutTable[lutIndex];
+
+	// Coarse z. The bin the stub is going to be put in, in the memory
+	int bin = lutValue >> nbitsfinerz; // 3 bits, i.e. max 8 bins within each VM
+
+	if (negDisk) bin += 1 << (nbitsbin-1); // The upper half of the bins are for negative disks
+	
+	auto ivm = phicorr.range(phicorr.length() - nbitsall - 1, phicorr.length() - (nbitsall + vmbits)); //get the phi bits that corresponds to the old vms
+	slot = ivm * (1 << nbitsbin) + bin;
+
+	// Set rzfine, i.e. the r/z bits within a coarse r/z region
+	auto rzfine = lutValue & ((1 << nbitsfinerz) - 1); // the 3 LSB as rzfine
+	stub.setFineZ(rzfine);
+
+	assert(rzfine >= 0);
+	
+	return stub;
+};
+
+/////////////////////////////////
+// Main function
+
+// inType will be BARRELPS, BARREL2S, or DISKPS (since methods not defined for 'DISK' AllStubs, despite both DISKPS and 2S contained in each memory)
+// and outType will be BARRELPS, BARREL2S, or DISK (since both DISKPS and 2S contained in each VMStubME memory)
+template<int layer, int disk, regionType inType, regionType outType, int rzSizeME, int phiRegSize>
+void VMSMERouter(const BXType bx, BXType& bx_o,
+		// LUTs
+		const int METable[],
+		const int phiCorrTable[],
+		// Input memories
+		AllStub<inType>& allStub,
+		// Output memories
+		VMStubMEMemoryCM<outType, rzSizeME, phiRegSize, kNMatchEngines> *memoryME,
+		AllStubMemory<outType> allStubsCopy,
+		// Index of AllStub processed (used for setting 'index' in VMStub)
+		unsigned int index,
+		// Array to count how many VMStubs written in each slot
+		ap_uint<5> addrCountME[1 << (rzSizeME + phiRegSize)],
+		// Bool if valid stub
+		bool valid) {
```

Comment:
@mlarson02 I think a few consts can be sprinked in here as well. Also, I think `allStubsCopy` needs to be a reference? Passing it by value means that any changes you make to it, such as on line 183, will be lost when the function returns.
```suggestion
		const AllStub<inType>& allStub,
		// Output memories
		VMStubMEMemoryCM<outType, rzSizeME, phiRegSize, kNMatchEngines> * const memoryME,
		AllStubMemory<outType>& allStubsCopy,
		// Index of AllStub processed (used for setting 'index' in VMStub)
		const unsigned int index,
		// Array to count how many VMStubs written in each slot
		ap_uint<5> addrCountME[1 << (rzSizeME + phiRegSize)],
		// Bool if valid stub
		const bool valid) {
```